posthog/models/hog_function_template.py

```diff
@@ -0,0 +1,317 @@
+from django.db import models
+from django.db.models import Subquery, OuterRef, Max
+
+from typing import Literal, Optional, cast
+import hashlib
+import structlog
+import dataclasses
+
+from posthog.models.utils import UUIDModel
+from posthog.cdp.templates.hog_function_template import (
+    HogFunctionTemplateType,
+    HogFunctionTemplateKind,
+    HogFunctionTemplate as HogFunctionTemplateDTO,
+    HogFunctionMapping,
+    HogFunctionMappingTemplate,
+)
+
+logger = structlog.get_logger(__name__)
+
+
+class HogFunctionTemplate(UUIDModel):
+    """
+    Django model for storing HogFunction templates in the database.
+    This model replaces the in-memory storage of templates and enables sha versioning,
+    efficient storage, and easier updates for HogFunction templates.
+    """
+
+    # Core Template Information
+    template_id = models.CharField(max_length=255, db_index=True)
+    sha = models.CharField(max_length=100, db_index=True)
+    name = models.CharField(max_length=400)
+    description = models.TextField(blank=True, null=True)
+
+    # Core Template Content
+    code = models.TextField()
+    source = models.CharField(max_length=20, default="hog")  # "hog" or "javascript"
```

Comment:
Ooh this threw me. I consider "source" and "code" to be two words for the same thing `source-code`
Can we do `code` for the code and `code_language` or something for the kind? I could imagine a word like `source` also meaning "where did this code come from" like a github url or something, hence me thinking `code` is the best fit